quesma/optimize/trunc_date.go

```diff
@@ -0,0 +1,253 @@
+// Copyright Quesma, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+package optimize
+
+import (
+	"fmt"
+	"quesma/model"
+	"strings"
+	"time"
+)
+
+type truncateDateVisitor struct {
+	truncateTo time.Duration
+	truncated  bool
+}
+
+func (v *truncateDateVisitor) visitChildren(args []model.Expr) []model.Expr {
+	var newArgs []model.Expr
+	for _, arg := range args {
+		if arg != nil {
+			newArgs = append(newArgs, arg.Accept(v).(model.Expr))
+		}
+	}
+	return newArgs
+}
+
+func (v *truncateDateVisitor) processDateNode(e model.Expr) (string, string, bool) {
+
+	if fn, ok := e.(model.FunctionExpr); ok {
+
+		if fn.Name == "parseDateTime64BestEffort" {
+
+			if len(fn.Args) == 1 {
+
+				//"2024-06-04T13:08:53.675Z" -> "2024-06-04T13:08:00.000Z"
+
+				if date, ok := fn.Args[0].(model.LiteralExpr); ok {
+					if dateStr, ok := date.Value.(string); ok {
+
+						dateStr = strings.Trim(dateStr, "'")
+						return fn.Name, dateStr, true
+					}
+				}
+			}
+		}
+	}
+	return "", "", false
+}
+
+type dateComparisonNode struct {
+	column string
+	op     string
+	fn     string
+	date   string
+}
+
+func (v *truncateDateVisitor) processDateComparisonNode(e model.Expr) *dateComparisonNode {
+
+	if op, ok := e.(model.InfixExpr); ok && (op.Op == ">" || op.Op == "<" || op.Op == ">=" || op.Op == "<=") {
+		if column, ok := op.Left.(model.ColumnRef); ok {
+			if fn, date, ok := v.processDateNode(op.Right); ok {
+				return &dateComparisonNode{column: column.ColumnName, op: op.Op, fn: fn, date: date}
+			}
+		}
+	}
+	return nil
+}
+
+func (v *truncateDateVisitor) compare(column, op string, right model.Expr) model.Expr {
+	return model.NewInfixExpr(model.NewColumnRef(column), op, right)
+}
+
+func (v *truncateDateVisitor) date(fn, str string) model.Expr {
+	return model.NewFunction(fn, model.NewLiteral(fmt.Sprintf("'%s'", str)))
+}
+
+func (v *truncateDateVisitor) truncateDate(op string, t time.Time) string {
+
+	truncatedDate := t.Truncate(v.truncateTo)
+
+	if op == "<" || op == "<=" {
+		truncatedDate = truncatedDate.Add(v.truncateTo)
+	}
+
+	return truncatedDate.Format(time.RFC3339)
+}
+
+func (v *truncateDateVisitor) VisitLiteral(e model.LiteralExpr) interface{} {
+	return e
+}
+
+func (v *truncateDateVisitor) VisitInfix(e model.InfixExpr) interface{} {
+
+	op := strings.ToLower(e.Op)
+	if op == "and" {
```

Comment:
I think that body of this if branch could be extracted as separate function and rewritten a bit to reverse conditions,
```
  if op != "and" {
     return
  }
```

Reply:
Yep. I've extracted it to a separate method.